### PR TITLE
ENH+BF: save --super-datasets, save subds, save [files] avoiding saving index etc

### LINKED ISSUE #472: helper to automate commits for the datasets (including super-datasets and its "children")

### PR DESCRIPTION
**Pull request by yarikoptic**
Closes #472

### CODE REVIEW

**Comment by bpoldrack** [comment]
To be honest: I don't really like it. I can see, how `--super-dataset` might be convenient sometimes, but I think it's strange to do things upwards and it doesn't really follow the idea of saving the state of a certain dataset, if that means to save the state of not contained datasets, too.
What is so troublesome about just saving the superdataset?

**Comment by mih** [comment]
FTR: I am feeling the same way.

**Comment by yarikoptic** [comment]
Your "sometimes" might be mine "often". Usecase - recrawl one openfmri dataset, and now need to issues two saves on top manually. For datasets deeper in hierarchy - even more times

**Comment by bpoldrack** [comment]
Well, then we may be should talk about how to make sure, that a `save TOP` goes down, the way you want it to, in order to have a single call only. That's something I could easily agree on.

**Comment by yarikoptic** [comment]
after I wrote some "other" reply, I thought we could indeed possibly use, I can live with
```save -d^``` which would then need logic for going through subdatasets until it hits PWD one.  Is that what you wanted @bpoldrack ?
it would need the same enhancement in deciding when to stop going up applied more generically though (which would benefit search which relies on -d^ ATM as well)

**Comment by yarikoptic** [comment]
to be complete, I would need to issue `save -d^ $PWD` and then save should get a logic, if path to a subdataset provided, to save all leading to it subdatasets (in decreasing depth order).  Then it would be a generic approach.  like it?

**Comment by bpoldrack** [comment]
Well, kind of. I feel, that there's something I don't get for now.  Generally, that's the direction I feel is right, but:

That's not exactly what I meant. What probably is unclear to me is why you need to specify a certain subdataset, to step down to or to stop the recursion at. Let's say you have a dataset `TOP` and create or modify a subdataset `SUB` within `TOP`'s tree. What you want to do now, is to "register" `SUB` as a submodule (or update the commit, the submodule is referring to or whatever). The only thing you need to do from my point of view is saving `TOP`. So, I think it should just be `save TOP` (or `save -d^`, of course). Why would you need to specify what subdataset of `TOP` was modified? `save` already does figure this out. `git` knows where there are modifications, no? Why do we need additional logic here? Nothing else was changed in your usecase and even if so - why would you want to not save that? I'm not entirely sure, what I do not understand, but somehow it is not clear to me, what's the issue.

**Comment by yarikoptic** [comment]
I need such logic if/when there are also other changes in super-dataset and/or some other sub-dataset(s) which I don't want to save ATM.  So overall super-dataset is already "dirty" with some other changes, and I need to save only that particular sub-dataset and propagate this "saving" all the way up so I could publish in a current state.  Example:  I am working/fixing crawling in super/sub1/subsub1, so that super has sub1 dirty with some new commits within subsub1.  But then I recrawled super/sub2/subsub2 and need to save and publish that.  ATM I would need to save manually super/sub2/subsub2, then super/sub2/ and then within super itself pointing specifically at sub2.  I.e. lots of navigation and 'save' invocations I need to redo.  I cannot just save `super` with everything changed.

**Comment by bpoldrack** [comment]
Ok. Then indeed it generally should be `save -d super sub2/subsub2` like it would be in case it was about a modified file rather than a modified subdataset.

**Comment by mih** [comment]
`save -d^ $PWD`  seems like a good solution. It is inline with the general behavior of `save` and the traversal rule makes sense.

Current implementation in PR is still the initial approach, it seems -- will not comment on it.

**Comment by yarikoptic** [comment]
:-( when I got to change it toward "save -d^ $PWD" I have realized that it would then require ugly constructs in Python code because in Python API I see us/users primarily operate via dataset methods (and that is what we do usually in the tests), so we don't need to cd to that directory,  where in cmdline usually we work with that dataset working directory.  So in Python API "dataset" is already specified since I would just call `subds.save(...)`.  I would then need to do something as ugly as `subds.get_superdataset(topmost=True).save(subds.path)` to achieve the same as `save -d^ $PWD`.

Moreover, specifying $PWD could be also 'detrimental' somewhat if e.g. I am in a sub-directory of a dataset, so I would need first to get to its root (within BIDS dataset could be 3 levels down or more) to issue such a command, since otherwise it would mean a different thing... 

And anyways, if we are to implement support for `save -d^ sub/sub/path` -- it would carry the same "going up" first (as part of `-d^`), then collecting where to go down (by matching among subdatasets), then issuing saves while going up again through those collected paths (since we want to record changed states in kids first)... I am even scared to think how it would work out if there are some fancy symlinks mixed in ;)  so actually for now (or forever) I would prefer to stay with current implementation. As you can see -- it is quite lean and straightforward. It relies on the same mechanism of `get_superdataset` as used by `-d^` so whatever we decide about boundaries between datasets and repos -- it will be ready as long as that method adheres to new desired logic.

What could may be a better is the name -- but I could not see a better one, as e.g some antonym to 'recursive' as implying downwards.

**Comment by yarikoptic** [comment]
BTW, I needed it today again, even had to switch a branch just to get it available

**Comment by bpoldrack** [comment]
Well, I don't think `subds.get_superdataset(topmost=True).save(subds.path)`would be particularly ugly.
Moreover: While `Dataset.save()` might be the usual approach, you nevertheless could use `datalad.api.save(dataset="^", subds.path)` to achieve the same thing.

Regarding all the up and down traversal: How would that be different in case of `--super-dataset`? Whether you specify it via `-d^` or that switch - it still has to do the same thing, no?

**Comment by yarikoptic** [comment]
@bpoldrack -- thanks for python API examples -- indeed it might be not that ugly after all

> How would that be different in case of --super-dataset? 

it just does traversal up (that is what is implemented), while immediately calling out 'save' in every dataset it visits until it is on the top.

With `dataset="^"` it would need first to do traversal up, and then figure out all the breadcrumbs (i.e. sub-datasets paths) which lead to the `path` (probably by doing that traversal from the `path` again), take the ones which start from the root dataset.  It could be implemented by going down, but I think it would be inferior since it would ask for all submodules at each super-dataset "exploration".
Breadcrumbs collection could be facilitated if `get_superdataset` could get a parameter which would return all those breadcrumb paths, instead of just the final superdataset, but again -- it seems we wouldn't avoid traversal twice one way or another.

But I do think that eventually such implementation would be needed/useful so we could save while PWD in super-datasets.

**Comment by yarikoptic** [comment]
so this discussion brought me to #1024, and then it would indeed be somewhat nice to be able to say `save -d^` (probably without PATH since we could just would then use $PWD anyways, right?) as to save starting from direct super, or `save -d^*` for the ultimate top super-dataset.

**Comment by bpoldrack** [comment]
> (probably without PATH since we could just would then use $PWD anyways, right?)

Yes, makes sense and that's what I think it would result in.

**Comment by yarikoptic** [comment]
probably $PWD would be mandatory for such a case since otherwise we couldn't tell if that was just a regular `Dataset('^').save()` call ran from outside of the dataset (but might currently be under one of the subdatasets), which we allow (e.g. for Python API).  So PATH needs to be explicitly specified although that is somewhat complicates my life since if I am inside subdir of the subdataset I would be effectively instructing to save only changes in that directory, so I need to go up... yet thinking about it and looking at save code (and how it silently ignores `files` if `-a` is given ;) )

File: datalad/distribution/tests/test_dataset.py [hunk 1]
```diff
@@ -135,6 +135,7 @@ def test_is_installed(src, path):
     # submodule still not installed:
     subds = Dataset(opj(path, 'subm 1'))
     assert_false(subds.is_installed())
+    subds.create()
```
Comment (bpoldrack) [inline datalad/distribution/tests/test_dataset.py]: What is that about? This should not be necessary!

File: datalad/distribution/tests/test_get.py [hunk 2]
```diff
@@ -264,6 +264,7 @@ def test_get_recurse_subdatasets(src, path):
 def test_get_install_missing_subdataset(src, path):
 
     ds = install(path=path, source=src)
+    ds.create(force=True)  # force, to cause dataset initialization
```
Comment (bpoldrack) [inline datalad/distribution/tests/test_get.py]: Again: install is not sufficient? Why is this?

File: datalad/interface/tests/test_save.py [hunk 3]
```diff
@@ -113,3 +115,28 @@ def test_recursive_save(path):
...
+    create_tree(subsubds.path, {"testnew2": 'smth'})
+    assert subsubds.repo.dirty
+    # and indexed files didn't change
+    assert_equal(indexed_files, subds.repo.get_indexed_files())
+    subsubds.save(message="saving", super_datasets=True,
+                  auto_add_changes=True)
+    assert not subsubds.repo.dirty
+    # but its super should have got only the subsub saved
+    # not the file we created
+    assert subds.repo.dirty
```
Comment (bpoldrack) [inline datalad/interface/tests/test_save.py]: This does not test, what you want to be the result according to the comment. Just "dirty" is also true, if your `super_datasets=True` did nothing.

**Comment by bpoldrack** [comment]
I guess there are commits to come anyway, but just in case submitted a review. I'm aware that it may become invalid anyway.

File: datalad/distribution/tests/test_dataset.py (hunk 1 above)
Comment (yarikoptic) [inline datalad/distribution/tests/test_dataset.py]: Remember that we have 3 types of datasets! :-)  (see #1010)

File: datalad/distribution/tests/test_get.py (hunk 2 above)
Comment (yarikoptic) [inline datalad/distribution/tests/test_get.py]: Remember that we have 3 types of dataset. Source path is pure git/annex - not a dataset

**Comment by yarikoptic** [comment]
Re Commits... They might come indeed but not sure when... Lots todo and if to come should be on top of #1011 to reuse it's func... But moreover still feel that just a simple option with clean implementation might be of value even if we support alternative invocation via path

**Comment by yarikoptic** [comment]
grr.. PY3 failures on travis but can't reproduce locally -- for some reason on travis gitrepo refuses to provide hexsha of the current commit... strange.  If someone has a clue -- would be appreciated.

File: datalad/distribution/dataset.py [hunk 4]
```diff
@@ -359,19 +359,21 @@ def is_installed(self):
...
 
-    def get_superdataset(self, topmost=False):
+    def get_superdataset(self, datalad_only=False, topmost=False):
         """Get the dataset's superdataset
 
         Parameters
         ----------
+        datalad_only : bool, optional
+          Either to consider only "datalad datasets" (with non-None
+          id), or (if False, which is default) - any git repository
```
Comment (yarikoptic) [inline datalad/distribution/dataset.py]: relates to #1010

**Comment by yarikoptic** [comment]
HA -- tests now passed. hip hip -- hoorray. BTW -- current save implementation is very inefficient happen there is a reasonable number of files provided to its call, so we should eventually RF it

**Pull request merged**